### PR TITLE
Driver stream math refactor

### PR DESCRIPTION
**Pull request by cnuernber**
Huge merge.

This is the landing of a lot of careful work to build a stand-alone math library that cortex uses for its layers as opposed to having a lot of math tied into cortex itself.

Also, batch normalization now has a spatial component.

### CODE REVIEW

**Comment by mikera** [comment]
Hi @cnuernber I haven't read all of this code but it looks like you may be reinventing some parts of `core.matrix` unnecessarily. 

For example, why invent your own `gemm` protocols for example when core.matrix already has them in 
`clojure.core.matrix.blas1. See: https://github.com/mikera/core.matrix/blob/develop/src/main/clojure/clojure/core/matrix/blas.cljc

**Comment by cnuernber** [comment]
My intention is to implement enough of the core.matrix protocols and make the tensors library a partial citizen at some point in the future.  This is where you may be helpful; I already implemented some of the core.matrix protocols in various places in the tensors.clj file:

https://github.com/thinktopic/cortex/blob/driver-stream-math-refactor/src/cortex/tensor.clj#L248

Main points:
 
.  This is not meant to be a competitor to core.matrix but rather to fit into it's ecosystem as a more complex-to-use backend with functions and abstractions to explicitly allow easy integration with cudnn type libraries and of course things like multiple gpu and seamless opencv support. 

The library is carefully designed to provide an abstraction layer to ensure that backends do not duplicate implementation code and that clients can use the precisely same code and run it on the cpu, cuda, and opencv.  In order to achieve this end, I added primitives that do not exist in core.matrix.  These include:

*  The stream abstraction that allows an implementation with identical characteristics across the cpu, cuda, and opencv.

* The device abstraction so that clients can use multiple gpus and the library has knowledge about how to move data between devices along with code to detect possible conflicts due to this ability. 

* unified buffer abstraction which explicitly separates the heavy buffer from the descriptor to allow clients to do something like allocate one large buffer and create many tensors from one buffer at various offsets.  This allows things like Torch's ability to allocate all parameters from exactly one buffer and thus the optimize call to adam or whatever only happens exactly once per batch instead of once per parameter buffer per batch.  This implies allowing allocation and offsetting to be supported in a backend independent way. 

*   Explicit support of multiple primitive datatypes.  Unfortunately this is a hard requirement due to the vast performance differences on the gpu when using float vs double vs half-float datatypes.  It also opens the door to more exotic things like an image processing layer which uses the primitives from nvidia's npp library in cuda but uses the image processing capabilities of imagez and by extension think.image.  This includes a marshalling system by the way so that one can transform data between the datatypes seamlessly with an assign! call.

**Comment by cnuernber** [comment]
Another thing that would be extremely helpful would be unit tests for broadcasting and slicing that cover some amount of the python API for these operations.  I know the vectorz implementation supported some or all of this and the index system abstraction in cortex.tensors is explicitly designed to allow some amount of these types of abstractions on the GPU.  This is good direction that could prove very fruitful.

**Comment by mikera** [comment]
Do you have a design doc?

I'm not entirely clear abolut the idea of the stream abstraction you have in mind, but core.matrix already allows transparent operations across CPU, GPU, OpenCV etc. check out vectorz-opencl (https://github.com/mikera/vectorz-opencl) example which is a working OpenCL GPU implementation that is 100% compatible with core.matrix (and shares base types with vectorz, so CPU<->GPU operations can take advantage of this for performance)

A device abstraction *might* belong in core.matrix providing we can be clear exactly how it works. It sounds like this could be a extension of the current way that we specify different implementations for array construction in core.matrix (currently using keywords, but could support e.g. maps with device data)

There is something like a unified buffer abstrtaction already, in the sense that you can use `as-vector` and `subvector` to access arbitrary chunks of large arrays. 

Multiple data types is also fine with core.matrix. There's absolutely nothing in core.matrix itself that enforces any restriction on element types. vectorz happens to use doubles, but that's an implementation detail.

Regarding broadcasting, we already support broadcasting that covers most of what numpy does, and where we don't (for good reasons, I think numpy has design flaws here) then it is easy to get the same effect with explicit use of `reshape` and/or `broadcast`. These have reasonable test coverage in the compliance suite and in implementation, though of course I agree we can always improve test coverage ;-)

My overall impression is that `core.matrix` already supports most of what you want, and if there is a very specific thing missing then happy to add it (with usual caveats: as long as it has a sane API, is consistent with the rest of the design and generally within scope etc.)

**Comment by cnuernber** [comment]
I don't have time to review that and the opencv backend just an experiment.

I do not believe that any core.matrix backend supports enough functionality to implement cortex in it and if you want to prove me wrong then write the code that does so.  It should not be too much effort to get some subset of cortex working on cpu and gpu.  Come back with a branch of working code please :).

**Comment by mikera** [comment]
The original cortex (which as you know I mostly wrote along with @rosejn ) was fully working and implemented layers as fairly simple wrappers over a small number of core.matrix operations. I didn't run into anything that might be a problem.

See e.g. https://github.com/thinktopic/cortex/commit/186cec3722c83c789e8f99d01ea0680ed50b43b1 where I refactored a bunch of superfluous code just to use the core.matrix blas operations.

Happy to work with you to ensure that core.matrix covers everything you need (it should, it was designed from the start to support projects like this) but I think we need some kind of strategy / design doc. 

Could you outline the "stream math" design in a few notes so that I can cross check and get a core.matrix implementation up and running (or at least, point me to where such docs exist if they are already in place)

**Comment by cnuernber** [comment]
The original cortex was redesigned quite heavily and the current cortex has both a lot more capabilities and an enormous amount more performance.  Blas is just a very small piece of the puzzle.

Mike, the simplest way is to implement a backend for cortex using core.matrix.  I don't see why this is such a difficult thing to do.  Then over time people can choose which to use based on needs, capabilities and performance requirements.

The place to start doing a backend is the compute binding layer:

https://github.com/thinktopic/cortex/blob/master/src/cortex/compute/driver.clj

Then move into building out the backend.  A partial implementation of the tensor layer will give you the linear layer and such.  

At this point we can evaluate the differences between the two systems once we have concrete implementations.  It will then be simple to see what pieces need to move forward to combine them and how.

**Comment by rosejn** [comment]
In Cuda you load data onto the card, and then you execute functions (kernels) and pass them pointers to your on GPU buffers along with function args.  (Probably the same as what you did in OpenCL.)  In order to enable many simultaneous tasks, and to hide memory latencies, you issue kernel calls on a stream, which has an integer ID.  Stream computations can run interleaved and in parallel.  I think what Chris is primarily referring to is that generic math calls that don't allow us to specify a stream ID won't let us effectively use most of the card.

In terms of the data types, although core.matrix doesn't specify concrete types, there isn't a non double implementation to use.  At least that I know of.  The tensors that Chris shifted to allow us to allocate bytes, floats, and doubles on both the CPU and GPU, and then efficiently DMA copy between them.  Part of the issue here is probably that we would have to first copy (with a type conversion) from a core.matrix vectorz double array into one of these memory locked CPU arrays of floats, and then do the copy onto the GPU.

There are probably more details, but at least from memory those are some of the possible concerns that come up.   It very well might be feasible though to find solutions for these issues.  For example by having the tensors we are using implement some key core.matrix protocols, and then maybe to use a binding or something like that to handle the stream ids.  That could get tricky though, and I don't think any of us have spent the time to figure out what implications it would have.

Hopefully that helps a bit.

**Comment by mikera** [comment]
@rosejn yes that's exactly what I did with OpenCL. Base data structure (https://github.com/mikera/vectorz-opencl/blob/master/src/main/java/mikera/vectorz/jocl/JoclVector.java) is just a wrapper over an OpenCL buffer. Other array types wrap these with strided access into these buffers. Works pretty well, the key point is that these details aren't necessary for the user to know, just regular core.matrix calls work (obviously the user can dive into implementation specific stuff if they want).

So basically what I'd like to see in Cortex is clean implementation-independent abstractions with several good solid implementations underneath. That is after all part of the overall rationale for Cortex. 

Maybe I'm missing something, but I'm not quite understanding why the same approach I used for vectorz-opencl couldn't also wrap things like Cuda stream handling etc. As far as I see it, the "tensor" defined in cortex/src/cortex/tensor.clj is fundamentally an array type that wraps device, stream and index handling, which makes sense for Cuda but might not for other backends. And then presumably you need to pass through all the different maths operations via the driver (if I understand the current design correctly) which will mean duplicating pretty much the whole core.matrix API.

It therefore seems to me like the layering is wrong and/or implementation is getting mixed up with abstractions. Ideally you would have a "tensor" abstraction that is independent of backends, but currently it seems like this has various leaky abstractions that depends on a lot of Cuda specific stuff. I completely understand that Cuda is your current priority so understand why it looks this way at the moment, but my experience with `core.matrix` (and Clojure in general) is that if you don't get these things right then the abstractions will not work well for alternative implementations.

The design I propose would be:

1. A "Tensor" abstraction (core.matrix should work for this, if it doesn't cover everything needed I will add it ;-) )
2. A "Driver" abstraction (including operations for creation / allocation / population of tensors)
3. A Cuda tensor implementation that has implementations for all the required tensor operations (add, abs, gemm etc.). This could be a core.matrix implementation and still have knowledge of stream IDs, roughly like the current tensor implementation in Cortex.
4. A Cuda implementation of the Driver abstraction that knows how to create Cuda-specific tensors (including appropriate usage of streams, indexing strategies, data types etc.)
5. A generic driver implementation that just uses core.matrix without needing knowledge of the underlying core.matrix tensor implementation. This would enable e.g. Vectorz to be used directly for JVM-only usage.
6. Optional other implementations (who knows what funky hardware or libraries appear in the future...)

This way you get the best of both worlds: you can have all the Cuda-specific operations and performance enhancements in the Cuda implementation without leaking these details into the higher level APIs. This is certainly what I originally had in mind for cortex. 

I'm certainly willing to put in work to make this happen (and I can definitely sort out all the core.matrix side) but I also need to be sure this is aligned with your strategy before I put significant effort into it. @cnuernber it looks like you are doing most of the work on the Cuda tensor based parts so you'd need to be willing to help refactor if we agree to go in this direction. 

To be very clear, I think Cortex should be a **general purpose Clojure ML library with great Cuda support as well as good alternative implementations**, *not* a Cuda-focused library with other implementations as second class citizens that have to pretend to be Cuda in some way. @cnuernber  to respond to your quite reasonable suggestion of writing a core.matrix backend, I am very happy to do this but only once we have good abstractions in place, otherwise I feel it would just be creating a lot of difficult to maintain boilerplate.

I hope this is clear, happy to elaborate further if not. Also please don't take anything I say as criticism, it is all honest feedback offered in the spirit of improving the architecture of Cortex.

What do you guys think? Where do you want to go with this?

**Comment by cnuernber** [comment]
There are a few things that aren't backend independent:

1.  Datatypes - support for multiple datatypes.  You have hardcoded vectorz to double; this is a mistake.  If you want to support floating point datatypes you have to, in your words, duplicate the entire core.matrix api.  The current cortex backend supports several datatypes with no duplication (or rather duplication that is maintained by macros or c++ templates).  Making datatypes a *part* of the api also allows support of for instance using the same math system to operate on images without casting each byte channel to a double precision number and then a marshalling (type converting) assignment to whatever datatype the network is running on.  There has been a *lot* of work making sure loading and working with arrays of various java primitive types is as fast as possible.  Not all operations (gemm for instance) are available across all datatypes.

2.  Separation of the buffer, memcpy, offsetting operations from the tensor design.  This allows me to create *any* tensor from any particular point in any buffer; so in essence it is more like you could create a densematrix type from your subvector or submatrix.

3.  The stream and device abstractions are specifically designed so that they would support opencv; they aren't cuda specific.  You can implement it with opencv primitives.  It gives a backend-independent way for the library to control the execution context so that we can write something like the batch upload process once and have it work across all backends.  We can also parallelize the execution of various components of the graph or the optimization system in backend independent ways.  

4.  The tensor design *is* carefully built so that backend implementations can be as minimal and so they have uniform and testable characteristics.  So it would make more sense to put the core.matrix protocols on top of the tensor abstraction because it will make implementing a complete opencv backend a lot less work *and* it will behave in a completely uniform manner compared to the cpu and cuda backends.  It also means that building a c-backend isn't too much effort (in that case same javacpp base datatypes as the cuda backend).

5.  There is a distinct breakdown of operations that need a CAS (compare-and-set) instruction and ones that do not.  The tensor math interface clearly differentiates between things that will need different kernels and things that will not regardless of opencv or cuda.  In this way we breakdown, for instance, binary operations that require three tensors (destination and the two operands) from operations that require only two and thus imply a cas operation (because the destination is being written to and read from).  The tensor library takes care of this breakdown so the backends do not need to.  So it forms a far more efficient platform to implement different backends that does any core.matrix system because it takes care of a lot of the machining around this.  

6.  There is no need for a vectorz backend.  This is an unnecessary distraction; we have yet to see a platform where the platform blas libraries do not load and given that the vectorz gemm is on the order of 100-1000 times slower than platform gemm...it seems pointless to me.

7.  Things like submatrix, subvector are taken care of in the library and described via dimensions.  This means backends do not need to do this removing a major source of bugs.  In core.matrix, every single backend has to reimplement major bug prone subsystems because specifically no backend does exactly what tensors.clj does; separate the buffer management from the description (dimensions) thus leading to a large amount of duplication.  So the implementation pathway for a new backend to cortex is shorter and much more likely to be correct than a new core.matrix backend; even if you take everything else described above into account.

8.  Multi-device support.  Being able to distribute load across multiple streams and multiple devices is key especially now that we are working on large images consistently.

**Comment by mikera** [comment]
Thanks @cnuernber for your detailed comments! Quick answers to all your points below.

1. This is something simply that the Driver abstraction+implementation should do. Specify in the driver API what type of data you want to be used when arrays are constructed. core.matrix itself will support any data type (there is even an implementation for complex numbers!). Obviously you may need to throw an exception if you ask an implementation to use a data type it doesn't support, but that's a user error anyway.
2. core.matrix API already supports this - subvector and reshape will do what you want. I suggest not reinventing the wheel here?
3. Obviously good if the Cortex Tensor supports OpenCV as well. Then it may make a decent array implementation for Cuda and OpenCV. I still don't think Tensor API users need to know about this.
4. I agree with putting the core.matrix protocols on top of the Tensor datatype, if the Tensor itself is used as the way to implement Cuda / OpenCV computation. It then de-facto becomes the "3. Cuda tensor implementation" that I mentioned in my previous list (and maybe also for OpenCV). Obviously a big bonus then is that any generic code that expects core.matrix arrays will work with any Tensors.
5. That's something I would suggest handling in the Driver abstraction to the extent that users care, but it seems mostly like an implementation detail. As for operations with a destination and two operands, there are several in core.matrix already (e.g. `add-outer-product!`), but it would be easy to add more if they are needed.
6. I want to be able run stuff in pure JVM code. That is a definite need for some of my use cases. Inability to do so would probably be a deal-breaker with Cortex for me (and is in fact the main reason I haven't used Cortex so much recently).
7. I think that's a design mistake. The whole point of core.matrix is that implementation can *optionally* implement such stuff efficiently. If you do it in a higher abstraction layer, you won't be able to take advantage of implementation specific operations. e.g. asking an implementation to return a subvector is much better than creating a subvector in the Tensor implementation and then..... doing what? indexed access into the underlying implementation? That's terrible for performance.
8. Again, belongs in the Driver interface+implementation, not baked into the Tensor API

**Comment by cnuernber** [comment]
1.  You missed the point.  The existing single backend supports all the datatypes efficiently (both in code size and implementation).  There is no core.matrix implementation that does this and you hardcoded all of vectorz to double along with building a superficial and pointless class hierarchy.

2.  reshape is copying thus making using one buffer for all parameters that gets optimized impossible.  You can see this is in fact impossible in that the vectorz implementation of dense matrix expects to own the entire array; it doesn't use both an array and an offset.  As far as that goes it is a design mistake to have a densematrix class and a submatrix class; these should only be differentiated by the shape and stride attached to the combination of data and offset.

3.  Sorry, I meant openCL.  The stream abstraction can be implemented with primitives in opencl; I already researched this and thought it through quite a bit.

4.  Not sure your point here.

5.  You missed the point.  First, it makes concrete implementations easier to differentiate between when the destination is being used as one of the operands because of the existence of CAS (compare-and-set) operation instead of requiring every implementation to do this at its level.

6. I don't believe this is necessary for any use case.  I have never seen netlib fail to load.  Concrete details here please.  This sounds like making up a need.

7. The tensor design makes things like that an offset operation long with setting up a dimension; totally handled in upper level code efficiently and the lower level code doesn't need to do it.  It is a far more efficient design in terms of separating concerns between what each backend needs to implement and what the tensor library can handle in a uniform way.  So I am fairly sure your original design in core.matrix is a design mistake.

8.  It isn't baked into the tensor api but a backend of cortex needs to be able to expose these things which the current one does.  That abstraction works into the stream and buffer abstraction (buffers have a device).  So all these concerns are unified in that it is important that the library knows if buffer objects are on the same device or not; again this is handled uniformly in the clojure code at the tensor.clj level so the cuda implementation for instance never needs to check something like that.  By extension an opencl implementation would not either.  So I have removed a lot of the concerns that happen if you attempt to implement multiple backends by making these types of assumptions in the tensor library.

The breakdown of core.matrix to the various backends is inefficient both in terms of developer time and in terms of potential bugs.  I have broken the math that cortex uses down in such a way as to provide the absolute simplest possible breakdown that will work for cortex across multiple different backend types; something that is very much missing from the core.matrix implementations because precisely there isn't a tensor abstraction thus leaving the entire implementation to each backend independently.

This is evident from a simple lines of code counting of any core.matrix backend.  The entire cortex cpu math system is about 1000 lines all told and it supports everything above and multiple datatypes.  This leads to long term benefits in terms of readability, extensibility, and testability.  Furthermore precisely because it contains an implementation of the stream and device system complex and bug prone algorithms in cortex need to be written precisely once instead of twice leading further to an increase in developer velocity.   For instance adding a binary operation, let's say pow(x,y) will take around 20 lines including the cuda backend.  Adding a similar type operation to core.matrix including protocol to implementation is going to run *far* longer so the current cortex system is better designed for precisely this type of extension which leads to better velocity in the long run.

So fundamentally it isn't about using the core.matrix protocols or not; that is irrelevant and not possible for operations like batch-norm, convolution, and max pooling.  It is about maintaining a unified codebase that is the simplest to maintain over a long period of time as the needs of cortex change but that allows things like multiple GPU support, total freedom with buffer management, and easy to use concurrency primitives to control how the system is executing subsections of graphs.  This is the point of the compute abstraction and the tensors sit on top of that.

There is space for a cortex implementation that isn't built on the compute abstraction at all and this is where if you want to spend your time doing something you could.  Cortex has significant tests so correctness will be verified.  A pure core.matrix backend won't have the features or the performance of the  existing system which is precisely where there is not one.
